src/analysis/bounds.rs

```diff
@@ -172,18 +172,8 @@ impl Bounds {
             Type::Fundamental(Fundamental::Filename) => Some(AsRef(None)),
             Type::Fundamental(Fundamental::OsString)=> Some(AsRef(None)),
             Type::Fundamental(Fundamental::Utf8) if *nullable => Some(Into(Some('_'), None)),
-            Type::Class(..) if !*nullable => {
-                if env.class_hierarchy.subtypes(type_id).next().is_some() {
```

Comment:
This change breaks quite a bit now, e.g. all the functions in `gio` taking `Into<Option<Cancellable>>` will now take a `Into<Option<IsA<Cancellable>>>`. While not a problem per se, it requires bigger changes in the code generator.

The nullable code currently handles everything different from the other `to_glib` conversions. Instead of directly calling `x.to_glib_none().0` in the function call, it goes via `ToGlibStash`

```rust
let foo = foo.into(); // Into<Option<_>>
let foo = foo.to_glib_none(); // ToGlibStash
[...]
ffi::bar(foo.0);
```

This is inconsistent with how we generate everything else, also quite visible in the code by having an early exit and lots of special cases for this.

What we now would need is
```rust
let foo = foo.into(); // Into<Option<_>>
[...]
ffi::bar(foo.map(AsRef::as_ref).to_glib_none().0);
```

Compared to non-nullable cases this is only a small difference (and more consistent too)

```rust
ffi::bar(foo.as_ref().to_glib_none().0);
```

But I'm not entirely sure yet how to let the code generator do that.

I would leave the `x.into()` as a separate transformation step at the top of the function but somehow I need to distinguish the `as_ref()` transformation needed for `IsA<_>` between nullable and not nullable.

Also what worries me a bit is the `ToGlibPointer { explicit_target_type }` special case. I'm not sure when exactly that is used. But maybe it just works fine with the above changes.

@EPashkin any suggestions?

Reply:
Found a nice solution for this, which also cleans up some code.